### PR TITLE
Feature/pyomo tools

### PR DESCRIPTION
**Pull request by bwronska**
Collection of functions to assist pyomo models.

### CODE REVIEW

**Review by ggsdc — CHANGES_REQUESTED** [review]
I would move all the code inside the existing models module with just a file called pyomo so that we can call the functions like: `from mango.models.pyomo import is_feasible`

The tests then have to be included as well inside the modules test folder

**Comment by bwronska** [comment]
> I would move all the code inside the existing models module with just a file called pyomo so that we can call the functions like: `from mango.models.pyomo import is_feasible`
> 
> The tests then have to be included as well inside the modules test folder

Done

**Review by hugolarzabal — CHANGES_REQUESTED** [review]

File: mango/pyomo/pyomo_tools.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
+import pyomo.environ as pyo
+from pytups import SuperDict
+from mango.table import Table
+from mango.processing import write_excel, as_list, load_json, write_json, load_excel_light
```
[the remark this answers is not in the document]
Comment (bwronska) [reply mango/pyomo/pyomo_tools.py]: done

File: requirements.txt [hunk 2]
```diff
@@ -21,4 +21,6 @@ streamlit<=1.28.0
 tqdm<=4.66.1
 unidecode<=1.3.7
 urllib3<=2.0.7
-XlsxWriter<=3.1.9
\ No newline at end of file
+XlsxWriter<=3.1.9
+Pyomo<=6.6.2
+Pillow<=10.0.1
```
Comment (hugolarzabal) [inline requirements.txt:26]: why is Pillow added to requirements? I do not see it used anywhere.

Comment (ggsdc) [reply requirements.txt:26]: It is the dependency of a dependency so it is no problem. And the requirements file is only used internally not for the building of the library.

**Review by ggsdc — APPROVED** [review]
Everything seems alright to me. @hugolarzabal what do you think?

File: mango/models/pyomo.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    :param instance: model instance (created with create_instance)
+    :param opt: solver instance (created with solver factory)
+    :return:
+    """
+    opt._presolve(instance)
+    opt._write_soln_file(instance, filename)
+
+
+def variables_to_excel(model_solution, path, clean=True):
+    """
```
Comment (hugolarzabal) [inline mango/models/pyomo.py:138]: maybe we should set rounding as an argument of the function (with default value).
Same in other functions where rounding is used.

File: mango/models/pyomo.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    :param path: path of the directory where to save the file
+    :param clean: discard variables with values of 0.
+
+    :return: nothing
+    """
+    variables = {
+        k: var_to_table(v, keys=None, clean=clean, rounding=6)
+        for k, v in model_solution.component_map(ctype=pyo.Var).items()
+    }
+    write_excel_light(path + "variables.xlsx", variables)
```
Comment (hugolarzabal) [inline mango/models/pyomo.py:151]: Maybe we should add the name of the file as an argument with default name being "variables.xlsx"

Comment (hugolarzabal) [reply mango/models/pyomo.py:151]: same for other functions

File: mango/models/pyomo.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    :param instance: model instance
+    :param path: the path of the file to be loaded
+    :return: model instance
+    """
+    data = load_excel_light(path + "variables.xlsx")
+    return instance_from_dict(instance, data)
+
+
+def instance_from_json(instance, path):
```
Comment (hugolarzabal) [inline mango/models/pyomo.py:183]: we should add kwargs to be passed to load_json.
quite often, you need to specify encoding for load_json.

**Review by hugolarzabal — CHANGES_REQUESTED** [review]

File: mango/models/pyomo.py (hunk 5 above)
Comment (bwronska) [reply mango/models/pyomo.py]: corrected

File: mango/models/pyomo.py (hunk 4 above)
Comment (bwronska) [reply mango/models/pyomo.py]: corrected

File: mango/models/pyomo.py (hunk 3 above)
Comment (bwronska) [reply mango/models/pyomo.py:140]: corrected

**Review by hugolarzabal — APPROVED** [review]

**Review by ggsdc — APPROVED** [review]

**Pull request merged**